### ISSUE TITLE
How to use the Rest API to match Nvidia-smi's memory output

### ISSUE THREAD

**Issue by nma** [NVIDIA/nvidia-docker#220]
```
{
    "Devices": [
        {
            "Clocks": {
                "Cores": 1189,
                "Memory": 3304
            },
            "Memory": {
                "ECCErrors": {
                    "Global": null,
                    "L1Cache": null,
                    "L2Cache": null
                },
                "GlobalUsed": 9931
            },
            "PCI": {
                "BAR1Used": 2,
                "Throughput": {
                    "RX": 136,
                    "TX": 48
                }
            },
            "Power": 122,
            "Processes": [
                {
                    "MemoryUsed": 9929,
                    "Name": "python",
                    "PID": 6151
                }
            ],
            "Temperature": 81,
            "Utilization": {
                "Decoder": 0,
                "Encoder": 0,
                "GPU": 76,
                "Memory": 6
            }
        }
    ]
}
```

Hi we are trying to use this API to create a dashboard for our users so they avoid stepping on each other's jobs whilst scheduling work on the GPUs.

Though we noticed that the memory Utilization does not match what Nvidia-SMI returns, and the documentation is not clear. 

As the Utilization is not total gigabytes, nor is it a percentage.

How would we use the output to match nvidia-smi's memory number?

**Comment by flx42** [comment]
It's a percentage, and it should match the output of `nvidia-smi dmon`, is it not the case?

**Comment by juharris** [comment]
@flx42 Thanks, the API does seem to match `nvidia-smi dmon` on the machines that I checked.
Yet our concern in that `nvidia-smi` reports memory like: `23MiB / 12285MiB` and `10026MiB / 12287MiB` whereas `nvidia-smi dmon` reports `1` and `0` respectively under `mem`.

Here's some output from one machine:
```bash
$ nvidia-smi dmon
# gpu   pwr  temp    sm   mem   enc   dec  mclk  pclk
# Idx     W     C     %     %     %     %   MHz   MHz
    0    30    52     0     1     0     0   405   135
    1   108    73    78     2     0     0  3304  1177
    0    30    52     0     1     0     0   405   135
    1   122    73    74     0     0     0  3304  1177
```

```bash
$ nvidia-smi
Fri Oct 14 16:54:19 2016
+------------------------------------------------------+
| NVIDIA-SMI 352.63     Driver Version: 352.63         |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX TIT...  Off  | 0000:01:00.0     Off |                  N/A |
| 22%   52C    P8    30W / 250W |     23MiB / 12285MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX TIT...  Off  | 0000:03:00.0     Off |                  N/A |
| 32%   73C    P2   105W / 250W |  10026MiB / 12287MiB |     78%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    1     11993    C   python                                       10000MiB |
+-----------------------------------------------------------------------------+
```

So maybe the docs here: https://github.com/NVIDIA/nvidia-docker/wiki/nvidia-docker-plugin#version-10 should be changed to clarify that.

**Comment by flx42** [comment]
Ah, I understand the confusion now, we are using the terminology of the `NVML` library, that's why.
For GPU/memory utilization, we are using function `nvmlDeviceGetUtilizationRates`, you can take a look at the NVML API in `/usr/local/cuda/include/nvml.h` to see the documentation of this function:
```
typedef struct nvmlUtilization_st  
{   
    unsigned int gpu;                //!< Percent of time over the past sample period during which one or more kernels was executing on the GPU    
    unsigned int memory;             //!< Percent of time over the past sample period during which global (device) memory was being read or written
} nvmlUtilization_t; 

nvmlReturn_t DECLDIR nvmlDeviceGetUtilizationRates(nvmlDevice_t device, nvmlUtilization_t *utilization);
```

So, memory utilization is not related to the amount of free memory on the GPU, it's related to the number of memory accesses made by your running program in the last sampling period. It tells you if your app is memory-intensive.

To get the amount of GPU memory being used, you want this line:
```
                "GlobalUsed": 9931
```

The total amount of memory for a GPU is static and can be retrieved with the `/gpu/info` call:
```
 curl -s http://localhost:3476/gpu/info/json | python -m json.tool
{
    "Devices": [
        {
            "Arch": "6.1",
            "CPUAffinity": 0,
            "Clocks": {
                "Cores": 1911,
                "Memory": 5005
            },
            "Cores": 2560,
            "Family": "Pascal",
            "Memory": {
                "Bandwidth": 320320,
                "Constant": 64,
                "ECC": false,
                "Global": 8104,
                "L2Cache": 2048,
                "Shared": 96
            },
            "Model": "GeForce GTX 1080",
            "PCI": {
                "BAR1": 256,
                "Bandwidth": 15760,
                "BusID": "0000:01:00.0"
            },
            "Path": "/dev/nvidia1",
            "Power": 180,
            "Topology": null,
            "UUID": "GPU-67d704a1-202d-4f62-f5f1-81186ba21e04"
        },
[...]
```

**Comment by juharris** [comment]
Ahh thanks! We should have noticed that number.

**Comment by flx42** [comment]
Glad to know it's solved, closing.

**Comment by victorhcm** [comment]
Hi @flx42, where can I find `/usr/local/cuda/include/nvml.h`. I checked my installation (CUDA 7.5), but it is absent. Do I have to download it from somewhere else?

**Comment by flx42** [comment]
@victorhcm it's now included in CUDA 8.0 (`cuda-nvml-dev` package on ubuntu).
Before 8.0, it was part of the "GDK":
https://github.com/NVIDIA/nvidia-docker/blob/master/Dockerfile.build#L17-L20
